File: anomaly_detector/src/main/java/psd/Transaction.java
package psd;

public class Transaction {
    private String transactionId;
    private String userId;
    private Long cardNumber;
    private Double value;
    private Long timestamp;
    private Location location;
    private Double balance;

    public Transaction(String transactionId, String userId, Long cardNumber, Double value, Long timestamp, Location location, Double balance) {
        this.transactionId = transactionId;
        this.userId = userId;
        this.cardNumber = cardNumber;
        this.value = value;
        this.timestamp = timestamp;
        this.location = location;
        this.balance = balance;
    }

    public Double getValue() {
        return value;
    }

    public Long getCardNumber() {
        return cardNumber;
    }

    public String getTransactionId() {
        return transactionId;
    }

    public String toString() {
        return "Transaction{" +
                "transactionId='" + transactionId + '\'' +
                ", userId='" + userId + '\'' +
                ", cardNumber=" + cardNumber +
                ", value=" + value +
                ", timestamp=" + timestamp +
                ", location=" + location +
                ", balance=" + balance +
                '}';
    }
}

File: transaction_generator/transaction.py
from dataclasses import dataclass

from card import Card
from location import Location
from user import User


@dataclass
class Transaction:
    transaction_id: str
    card: Card
    value: float
    timestamp: int
    location: Location

    def __str__(self):
        return f"{self.transaction_id} : {self.card} : {self.value} : {self.timestamp} : {self.location}"

    def to_dict(self):
        return {
            "transaction_id": self.transaction_id,
            "card": dict(self.card),
            "value": self.value,
            "timestamp": self.timestamp,
            "location": dict(self.location),
            "balance": self.card.balance,
        }

    def __iter__(self):
        yield "transaction_id", self.transaction_id
        yield "user_id", self.card.user.user_id
        yield "card_num", self.card.card_num
        yield "value", self.value
        yield "timestamp", self.timestamp
        yield "location", {"latitude": self.location.latitude, "longitude": self.location.longitude}
        yield "balance", self.card.balance

    @staticmethod
    def from_dict(transaction_dict):
        return Transaction(
            transaction_id=transaction_dict["transaction_id"],
            card=Card.from_dict(transaction_dict["card"]),
            value=transaction_dict["value"],
            timestamp=transaction_dict["timestamp"],
            location=Location.from_dict(transaction_dict["location"]),
        )
